ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/HeartbeatHandler.java

```diff
@@ -0,0 +1,445 @@
+     * If a previously discovered host has not received any heartbeat in this window, it is
+     * marked as dead.
+     * For all other hosts that received heartbeats, they are processed to determine whether the
+     * host is alive or dead based on how many consecutive heartbeats it missed.
+     * @param windowStart the start time in ms of the current window
+     * @param windowEnd the end time in ms of the current window
+     */
+    private void processHeartbeats(final long windowStart, final long windowEnd) {
+      // No heartbeats received -> mark all hosts as dead
+      if (receivedHeartbeats.isEmpty()) {
+        hostsStatus.forEach((host, status) -> {
+          if (!host.equals(localHostString)) {
+            status.setHostAlive(false);
+          }
+        });
+      }
+
+      //For previously discovered hosts, if they have not received any heartbeats, mark them dead
+      for (String host: hostsStatus.keySet()) {
+        if (!host.equals(localHostString) && (receivedHeartbeats.get(host) == null
+            || receivedHeartbeats.get(host).isEmpty())) {
+          hostsStatus.get(host).setHostAlive(false);
+        }
+      }
+
+      for (Entry<String, TreeMap<Long, HeartbeatInfo>> entry : receivedHeartbeats.entrySet()) {
+        final String host = entry.getKey();
+        final TreeMap<Long, HeartbeatInfo> heartbeats = entry.getValue();
+        synchronized (heartbeats) {
+          // 1. remove heartbeats older than window
+          heartbeats.headMap(windowStart).clear();
+          // 2. count consecutive missed heartbeats and mark as alive or dead
+          final  boolean isAlive = decideStatus(host, windowStart, windowEnd, heartbeats);
+          final HostStatusEntity status = hostsStatus.get(host);
+          status.setHostAlive(isAlive);
+          status.setLastStatusUpdateMs(windowEnd);
+        }
+      }
+    }
+
+    private boolean decideStatus(final String host, final long windowStart, final long windowEnd,
+                              final TreeMap<Long, HeartbeatInfo> heartbeats) {
+      long missedCount = 0;
+      long prev = windowStart;
+      // No heartbeat received in this window
+      if (heartbeats.isEmpty()) {
+        hostsStatus.get(host).setHostAlive(false);
+        return false;
+      }
+      // We want to count consecutive missed heartbeats and reset the count when we have received
+      // heartbeats. It's not enough to just count how many heartbeats we missed in the window as a
+      // host may have missed > THRESHOLD but not consecutive ones which doesn't constitute it
+      // as dead.
+      for (long ts : heartbeats.keySet()) {
+        //Don't count heartbeats after window end
+        if (ts >= windowEnd) {
+          continue;
+        }
+        if (ts - configs.heartbeatSendIntervalMs > prev) {
+          missedCount = (ts - prev - 1) / configs.heartbeatSendIntervalMs;
+        } else {
+          missedCount = 0;
```

Comment:
Just noting that this algorithm, will reset missedCount =0, as soon as it sees an timely heartbeat.. I think this is what we want it to do for now. worth adding a comment in code.

Reply:
I am not sure I follow but let me try to answer: If `heartbeatSendIntervalMS=10` and the last received heartbeat was at `ts=90` and windowEnd is at `ts=100`, then  `windowEnd-prev-1 = 9` and `missedCount =  9/10 = 0` which is correct since the window is exclusive of the end.  If  last received heartbeat was at `ts=89` then `missedCount =  12/10 = 1` which is also correct. I don't see when `missedCount` becomes zero erroneously.